src/query/api/v1/handler/graphite/find.go

```diff
@@ -57,52 +57,75 @@ func NewFindHandler(
 	}
 }
 
+func mergeTags(
+	noChildren *storage.CompleteTagsResult,
+	withChildren *storage.CompleteTagsResult,
+) (map[string]bool, error) {
+	// sanity check the case.
+	if noChildren.CompleteNameOnly {
+		return nil, errors.New("tags result with no children is completing name only")
+	}
+
+	if withChildren.CompleteNameOnly {
+		return nil, errors.New("tag result with children is completing name only")
+	}
+
+	mapLength := len(noChildren.CompletedTags) + len(withChildren.CompletedTags)
+	tagMap := make(map[string]bool, mapLength)
+
+	for _, tag := range noChildren.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = false
+		}
+	}
+
+	// NB: fine to overwrite any tags which were present in the `noChildren` map
+	for _, tag := range withChildren.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = true
+		}
+	}
+
+	return tagMap, nil
+}
+
 func (h *grahiteFindHandler) ServeHTTP(
 	w http.ResponseWriter,
 	r *http.Request,
 ) {
 	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
 	logger := logging.WithContext(ctx)
 	w.Header().Set("Content-Type", "application/json")
-	query, rErr := parseFindParamsToQuery(r)
+	noChildrenQuery, childrenQuery, raw, rErr := parseFindParamsToQueries(r)
 	if rErr != nil {
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
 		return
 	}
 
 	opts := storage.NewFetchOptions()
-	result, err := h.storage.FetchTags(ctx, query, opts)
+	noChildrenResult, err := h.storage.CompleteTags(ctx, noChildrenQuery, opts)
```

Comment:
sorry, no context - what's the difference between `noChildrenResult` and `childrenResult`

Reply:
Probably easiest with an example:

Query is: `query=foo.*`

This will actually call two Aggregate queries on the session:
`foo.*.*`, translated(-ish) to: `[ { __g0__ =~ foo } , { __g1__ =~ .* }, { __g2__ =~ .* } ]`, which is `withChildrenResult`
 and `foo.*`, translated(-ish) to: `[ { __g0__ =~ foo } , { __g1__ =~ .* }, { __g2__ !=~ .* } ]`, which is `noChildrenResult`
Where ` =~ ` is the regex operator, and ` !=~ ` is the regex exclusion operator.

Then these two are combined to determine which fields have children, and which are leaf nodes.

Hopefully this makes sense, will add a comment to this effect

Reply:
hm using your example of `query=foo.*`, why not just translate into a single aggregate query as follows: ```[ { __g0__ =~ foo } , { __g1__ =~ .* }, { __g2__ =~ .* } ]```

it's exactly the same as the union currently done in a single pass afaict

Reply:
actually, do you even need to use a regex operator for `__g0__`? that can be a straight up term operator

Reply:
Good idea; added an exact matcher case for alphanumeric segments but will look into properly detecting regex-able statements